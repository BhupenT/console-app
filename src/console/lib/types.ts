export interface Products {
  [key: string]: any;
}

export type DESC = string;
export type ASC = string;

export interface Sort {
  byField: string;
  sortOrder: DESC | ASC;
}
export interface File {
  name: string;
  type: string;
  sort: Sort;
}

export interface TargetFields {
  products: string;
  extraMeta: string;
}

export interface ExportProductsConfigSettings {
  productCatalogUrl: string;
  extraMetaUrl: string;
  file: File;
  targetFields: TargetFields;
}
