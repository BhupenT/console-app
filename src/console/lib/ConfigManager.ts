import { Injectable } from '@nestjs/common';
import * as fs from 'fs';
@Injectable()
export class ConfigManager {
  constructor(private readonly configPath: string) {}

  getConfig(className: string): Promise<any> {
    return new Promise((resolve, reject) => {
      fs.readFile(this.configPath, 'utf8', (err, data) => {
        if (err) {
          // no file present
          reject(false);
        } else {
          // file present already initialised
          resolve(JSON.parse(data)[className]);
        }
      });
    });
  }

  createConfig(
    path: string,
    content: any,
  ): Promise<{ saved: boolean; text: any } | NodeJS.ErrnoException> {
    return new Promise((resolve, reject) => {
      fs.writeFile(path, content, (err) => {
        if (err) {
          reject(err);
        }
        resolve({
          saved: true,
          text: content,
        });
      });
    });
  }
}
