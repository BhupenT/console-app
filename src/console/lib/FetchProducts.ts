import { HttpService } from '@nestjs/common';

export class FetchProducts {
  constructor(private httpService: HttpService) {}

  async fetchProductsInfo(url: string, fields = 'all'): Promise<any> {
    const response = await this.httpService.get(url).toPromise();
    let productsInfo = response;
    fields.split('.').forEach((field) => {
      productsInfo = productsInfo[field];
    });
    return productsInfo;
  }
}
