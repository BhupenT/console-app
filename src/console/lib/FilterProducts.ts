export class FilterProducts {
  private filteredProducts: any[] = [];

  constructor(private byField: string = 'video_count', products: any[] = []) {
    if (products) {
      products.forEach((product) => {
        this.addProduct(product);
      });
    }
  }

  getAllProducts(): { [key: string]: any }[] {
    return this.filteredProducts;
  }

  getProduct(productIndex: number): { [key: string]: any } {
    return this.filteredProducts[productIndex];
  }

  addProduct(product: { [key: string]: any }): boolean {
    const currentCount = this.filteredProducts.length; // get count before adding
    product[this.byField]
      ? this.filteredProducts.unshift(product)
      : this.filteredProducts.push(product);
    return this.filteredProducts.length > currentCount ? true : false; // return true or false if its added or not
  }

  modifyProduct(
    productIndex: number,
    key: string,
    value: any,
  ): { [key: string]: any } {
    const requestedProduct = this.getProduct(productIndex);
    requestedProduct[key] = value;
    this.filteredProducts[productIndex] = requestedProduct;
    return this.filteredProducts[productIndex];
  }
}
