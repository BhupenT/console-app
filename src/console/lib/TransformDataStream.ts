import { Transform, TransformCallback, TransformOptions } from 'stream';

export class TransformData extends Transform {
  constructor(
    options: TransformOptions,
    private addCallBack?: (data) => Promise<{ data: any }>,
  ) {
    super(options);
  }

  async _transform(
    chunk: any,
    encoding: BufferEncoding,
    callback: TransformCallback,
  ): Promise<void> {
    if (this.addCallBack) {
      const { data } = await this.addCallBack(chunk);
      chunk = { ...chunk, ...data };
    }
    this.push(chunk);
    callback();
  }
}
