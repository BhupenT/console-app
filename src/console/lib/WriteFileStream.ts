import { Writable, WritableOptions } from 'stream';
import { openSync, writeSync, closeSync } from 'fs';

type json = string;
type txt = string;
interface JsonOpts {
  open?: string;
  separator?: string;
  close?: string;
  itemCount: number;
}

export class WriteFileStream extends Writable {
  private counter = 1;
  options: JsonOpts;

  private open: number;
  constructor(
    public filePath: string,
    public fileType: json | txt,
    jsonOpts?: JsonOpts,
    options?: WritableOptions,
  ) {
    super(options);

    this.open = openSync(filePath, 'w');

    if (fileType === 'json' && !jsonOpts) {
      // adds default if not mentioned
      this.options = {
        open: '[\n',
        separator: ',\n',
        close: '\n]',
        itemCount: 0,
      };
    }
    if (jsonOpts) {
      this.options = {
        open: jsonOpts.open || '[\n',
        separator: jsonOpts.separator || ',\n',
        close: jsonOpts.close || '\n]',
        itemCount: jsonOpts.itemCount,
      };
    }
  }

  _write(
    chunk: any,
    encoding: BufferEncoding,
    callback: (error?: Error | null) => void,
  ): void {
    // console.log({ counter: this.counter, itemCount: this.options.itemCount });
    if ('json' === this.fileType) {
      switch (true) {
        case this.counter === 1: // adds open tag and separator
          this.writeToFile(chunk, true, true);
          break;

        case this.counter < this.options.itemCount: // adds separator only
          this.writeToFile(chunk, false, true);
          break;

        default:
          // adds nothing
          this.writeToFile(chunk, false, false);
          break;
      }
    } else {
      this.counter > 1
        ? this.writeToFile(chunk, false, true)
        : this.writeToFile(chunk, true, true);
    }

    this.counter++;
    callback();
  }

  _final(callback: (error?: Error | null) => void): void {
    if (this.options.close) {
      writeSync(this.open, this.options.close); // adds closing separator
    }
    closeSync(this.open);
    callback();
  }

  writeToFile(chunk: any, prepend?: boolean, append?: boolean): void {
    try {
      if (prepend && this.options.open) {
        writeSync(this.open, this.options.open);
      }
      writeSync(this.open, JSON.stringify(chunk, null, 2));
      if (append && this.options.separator) {
        writeSync(this.open, this.options.separator);
      }
    } catch (error) {
      closeSync(this.open);
    }
  }
}
