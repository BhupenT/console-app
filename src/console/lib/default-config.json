{
    "ExportProductService": {
        "productCatalogUrl": "https://eve.theiconic.com.au/catalog/products?gender=female&page=1&page_size=10&sort=popularity",
        "extraMetaUrl": "https://eve.theiconic.com.au/catalog/products/${data.sku}/videos",
        "file": {
            "name": "output.json",
            "type": "json",
            "sort": {
                "byField": "video_count",
                "sortOrder": "DESC"
            }
        } 
    }
}
