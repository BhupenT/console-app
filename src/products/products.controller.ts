import { Controller, Get, Param } from '@nestjs/common';
import { ProductsService } from './products.service';
import {Products} from './products.types';

@Controller('products')
export class ProductsController {
    constructor(private readonly productService: ProductsService) {}
    @Get()
    getProducts(): Products {
        return this.productService.getProducts()
    }

    @Get(':id')
    getProduct(@Param('id') id): {info: string, id: number} {
        return this.productService.getProduct(id);
    }
}
