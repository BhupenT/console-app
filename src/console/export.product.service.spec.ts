import { Test, TestingModule } from '@nestjs/testing';
import { ExportProductService } from './export.products.service';
import { HttpModule, HttpService } from '@nestjs/common';
import { ConfigManager } from './lib/ConfigManager';
import { FetchProducts } from './lib/FetchProducts';
import { TransformData } from './lib/TransformDataStream';
import { WriteFileStream } from './lib/WriteFileStream';
import { Readable } from 'stream';
import { readFileSync } from 'fs';

describe('ExportProductService', () => {
  let service: ExportProductService,
    configSettings: { [key: string]: any },
    products: Promise<any>,
    transformTestFile: any,
    writeTestFile: any;
  const Config: ConfigManager = new ConfigManager(
      `${process.cwd()}/iconic-cli.json`,
    ),
    Fetch: FetchProducts = new FetchProducts(new HttpService());

  const productsRecieved = [
    {
      name: 'sampleProduct',
      video_count: 0,
    },
    {
      name: 'dummyProduct',
      video_count: 0,
    },
    {
      name: 'sampleProduct2',
      video_count: 1,
    },
    {
      name: 'sampleProduct3',
      video_count: 1,
    },
  ];

  const productsToWrite = [
    {
      name: 'sampleProduct',
      video_count: 0,
    },
    {
      name: 'dummyProduct',
      video_count: 0,
    },
    {
      name: 'sampleProduct2',
      video_count: 1,
    },
    {
      name: 'sampleProduct3',
      video_count: 1,
    },
  ];

  const productsToTransformWrite = [
    {
      name: 'sampleProduct3',
      video_count: 1,
      video_previews: [
        {
          url: "'https://sample-test-video-mp3-sampleProduct3.com'",
          _links: {
            self: {
              href:
                'https://eve.theiconic.com.au/catalog/products/sampleProduct3/videos',
            },
          },
        },
      ],
    },
    {
      name: 'sampleProduct2',
      video_count: 1,
      video_previews: [
        {
          url: "'https://sample-test-video-mp3-sampleProduct2.com'",
          _links: {
            self: {
              href:
                'https://eve.theiconic.com.au/catalog/products/sampleProduct2/videos',
            },
          },
        },
      ],
    },
    {
      name: 'dummyProduct',
      video_count: 0,
      video_previews: [],
    },
    {
      name: 'sampleProduct',
      video_count: 0,
      video_previews: [],
    },
  ];

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ExportProductService],
      imports: [HttpModule],
    }).compile();

    (service = module.get<ExportProductService>(ExportProductService)),
      (configSettings = await Config.getConfig('ExportProductService')),
      service.setConfig(configSettings);

    products = await Fetch.fetchProductsInfo(
      configSettings.productCatalogUrl,
      configSettings.targetFields.products,
    );

    const writeStream = new WriteFileStream(
      'output-write-test.json',
      'json',
      { itemCount: 4 },
      {
        objectMode: true,
      },
    );

    const readableStream = new Readable({ objectMode: true });
    // push to the reable stream
    productsToWrite.forEach((product: { [key: string]: any }) => {
      readableStream.push(product);
    });

    readableStream.push(null); // push null to close the stream as its finished
    readableStream.pipe(writeStream);

    // push to the reable stream
    /* here using productsReceived array because this was sorted in above test as javascript
     * sort functions pushes to the parent
     */

    const writeTransformStream = new WriteFileStream(
      'output-transform-write-test.json',
      'json',
      { itemCount: 4 },
      {
        objectMode: true,
      },
    );
    const readableTransformStream = new Readable({ objectMode: true });

    productsToTransformWrite.forEach((product: { [key: string]: any }) => {
      readableTransformStream.push(product);
    });

    readableTransformStream.push(null); // push null to close the stream as its finished
    const transformStream = new TransformData(
      { objectMode: true },
      async (data) => {
        try {
          return data.video_count
            ? {
                data: {
                  video_previews: [
                    {
                      url: `'https://sample-test-video-mp3-${data.name}.com'`,
                      _links: {
                        self: {
                          href: `https://eve.theiconic.com.au/catalog/products/${data.name}/videos`,
                        },
                      },
                    },
                  ],
                },
              }
            : { data: { video_previews: [] } };
        } catch (error) {}
      },
    );
    readableTransformStream
      .pipe(transformStream) // transfrom data
      .pipe(writeTransformStream);
  });

  afterEach(async () => {
    (transformTestFile = await readFileSync(
      `${process.cwd()}/output-transform-write-test.json`,
      { encoding: 'utf8' },
    )),
      (writeTestFile = await readFileSync(
        `${process.cwd()}/output-write-test.json`,
        {
          encoding: 'utf8',
        },
      ));
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should fetch list of products', async () => {
    expect(JSON.stringify(await service.getProducts())).toMatch(
      JSON.stringify(products),
    );
  });

  it('should sort products by video_count by decending order', async () => {
    expect(await service.sortProductsBy(productsRecieved)).toMatchSnapshot();
  });

  it('should write in a json file', async () => {
    expect(writeTestFile).toMatchSnapshot();
  });

  it('should transform and then write in a json file', async () => {
    expect(transformTestFile).toMatchSnapshot();
  });
});
