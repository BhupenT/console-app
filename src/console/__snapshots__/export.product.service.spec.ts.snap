// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportProductService should sort products by video_count by decending order 1`] = `
Array [
  Object {
    "name": "sampleProduct3",
    "video_count": 1,
  },
  Object {
    "name": "sampleProduct2",
    "video_count": 1,
  },
  Object {
    "name": "dummyProduct",
    "video_count": 0,
  },
  Object {
    "name": "sampleProduct",
    "video_count": 0,
  },
]
`;

exports[`ExportProductService should transform and then write in a json file 1`] = `
"[
{
  \\"name\\": \\"sampleProduct3\\",
  \\"video_count\\": 1,
  \\"video_previews\\": [
    {
      \\"url\\": \\"'https://sample-test-video-mp3-sampleProduct3.com'\\",
      \\"_links\\": {
        \\"self\\": {
          \\"href\\": \\"https://eve.theiconic.com.au/catalog/products/sampleProduct3/videos\\"
        }
      }
    }
  ]
},
{
  \\"name\\": \\"sampleProduct2\\",
  \\"video_count\\": 1,
  \\"video_previews\\": [
    {
      \\"url\\": \\"'https://sample-test-video-mp3-sampleProduct2.com'\\",
      \\"_links\\": {
        \\"self\\": {
          \\"href\\": \\"https://eve.theiconic.com.au/catalog/products/sampleProduct2/videos\\"
        }
      }
    }
  ]
},
{
  \\"name\\": \\"dummyProduct\\",
  \\"video_count\\": 0,
  \\"video_previews\\": []
},
{
  \\"name\\": \\"sampleProduct\\",
  \\"video_count\\": 0,
  \\"video_previews\\": []
}
]"
`;

exports[`ExportProductService should write in a json file 1`] = `
"[
{
  \\"name\\": \\"sampleProduct\\",
  \\"video_count\\": 0
},
{
  \\"name\\": \\"dummyProduct\\",
  \\"video_count\\": 0
},
{
  \\"name\\": \\"sampleProduct2\\",
  \\"video_count\\": 1
},
{
  \\"name\\": \\"sampleProduct3\\",
  \\"video_count\\": 1
}
]"
`;
